File: core/trade_analyzer.py
import math
import time
from dataclasses import dataclass, field
from enum import Enum
from typing import List


class FeeTier(Enum):
    TIER1 = 1
    TIER2 = 2
    TIER3 = 3


@dataclass
class MarketImpactParams:
    eta: float          # Temporary impact coefficient
    gamma: float        # Permanent impact coefficient
    lambda_: float      # Risk aversion
    sigma: float        # Volatility
    alpha: float = 1.0  # Temporary impact exponent
    beta: float = 1.0   # Permanent impact exponent


class OrderSide(Enum):
    BUY = "BUY"
    SELL = "SELL"

@dataclass
class OrderLevel:
    price: float
    quantity: float

@dataclass
class OrderBook:
    asks: List[OrderLevel] = field(default_factory=list)
    bids: List[OrderLevel] = field(default_factory=list)

@dataclass
class MakerTakerInput:
    is_market_order: bool
    volatility: float
    order_book_depth: float
    time_since_last_trade: float


@dataclass
class MakerTakerResult:
    maker_prob: float
    taker_prob: float


@dataclass
class NetCostInput:
    usd_amount: float
    slippage: float
    fee: float
    market_impact: float


@dataclass
class TradeCostResult:
    slippage: float
    fee: float
    market_impact: float
    net_cost: float
    maker_taker_result: MakerTakerResult


class TradeAnalyzer:
    def __init__(self, order_book, exchange_name: str, fee_tier: FeeTier):
        self.order_book = order_book
        self.exchange_name = exchange_name
        self.fee_tier = fee_tier

    def get_order_book(self):
        return self.order_book

    def compute_spread(self) -> float:
        if not self.order_book.bids or not self.order_book.asks:
            return 0.0
        return self.order_book.asks[0].price - self.order_book.bids[0].price

    def compute_market_impact(self, usd_qty: float, params: MarketImpactParams) -> float:
        temp_impact = params.eta * usd_qty ** params.alpha
        perm_impact = params.gamma * usd_qty ** params.beta
        execution_risk = 0.5 * params.lambda_ ** 2 * params.sigma ** 2 * usd_qty ** 2
        return temp_impact + perm_impact + execution_risk


    def compute_slippage(self, usd_amount: float, volatility: float, spread: float, exchange: str) -> float:
        # Dummy linear regression coefficients for OKX
        beta0 = 0.01  # base slippage
        beta1 = 0.12  # log(order size)
        beta2 = 0.6   # volatility weight
        beta3 = 0.25  # spread weight

        # Customize for other exchanges if needed
        # if exchange == "BINANCE":
        #     beta0, beta1, beta2, beta3 = ...

        log_order = math.log(usd_amount)
        return beta0 + beta1 * log_order + beta2 * volatility + beta3 * spread

    @staticmethod
    def _sigmoid(x: float) -> float:
        return 1.0 / (1.0 + math.exp(-x))

    @staticmethod
    def get_taker_fee_percent(tier: FeeTier) -> float:
        if tier == FeeTier.TIER1:
            return 0.001
        elif tier == FeeTier.TIER2:
            return 0.0005
        elif tier == FeeTier.TIER3:
            return 0.0001
        return 0.001

    def compute_fee(self, usd_amount: float) -> float:
        fee_percent = self.get_taker_fee_percent(self.fee_tier)
        return usd_amount * fee_percent

    def compute_market_impact(self, usd_qty: float, params: MarketImpactParams) -> float:
        temporary_impact = params.eta * usd_qty
        permanent_impact = params.gamma * usd_qty
        execution_risk = 0.5 * params.lambda_ * (params.sigma ** 2) * (usd_qty ** 2)
        return temporary_impact + permanent_impact + execution_risk
    
    def compute_volatility(self, prices: List[float]) -> float:
        if len(prices) < 2:
            return 0.0
        log_returns = [math.log(prices[i] / prices[i - 1]) for i in range(1, len(prices))]
        mean = sum(log_returns) / len(log_returns)
        variance = sum((r - mean) ** 2 for r in log_returns) / len(log_returns)
        return math.sqrt(variance)


    def estimate_maker_taker(self, input_data: MakerTakerInput) -> MakerTakerResult:
        w_market = 2.0
        w_volatility = 1.5
        w_depth = -1.0
        w_time = -0.5

        score = 0.0
        score += w_market * (1.0 if input_data.is_market_order else 0.0)
        score += w_volatility * input_data.volatility
        score += w_depth * input_data.order_book_depth
        score += w_time * input_data.time_since_last_trade

        taker_prob = self._sigmoid(score)
        maker_prob = 1.0 - taker_prob

        return MakerTakerResult(maker_prob, taker_prob)

    def compute_net_cost(self, input_data: NetCostInput) -> float:
        return input_data.slippage + input_data.fee + input_data.market_impact

    def analyze(self, order_book, usd_amount: float, fee_tier: FeeTier,
                mi_params: MarketImpactParams, mk_input: MakerTakerInput,
                recent_prices: List[float], exchange_name: str) -> TradeCostResult:
        self.order_book = order_book
        self.fee_tier = fee_tier
        self.exchange_name = exchange_name

        spread = self.compute_spread()
        volatility = self.compute_volatility(recent_prices)
        slippage = self.compute_slippage(usd_amount, volatility, spread, exchange_name)
        fee = self.compute_fee(usd_amount)
        market_impact = self.compute_market_impact(usd_amount, mi_params)

        net_input = NetCostInput(usd_amount, slippage, fee, market_impact)
        net_cost = self.compute_net_cost(net_input)

        maker_taker_result = self.estimate_maker_taker(mk_input)

        return TradeCostResult(slippage, fee, market_impact, net_cost, maker_taker_result)

File: src/trade_analyzer.cpp
#include "trade_analyzer.hpp"
#include <cmath>
#include <numeric>
#include <stdexcept>

void LatencyTimer::start() {
    start_time = std::chrono::high_resolution_clock::now();
}

long long LatencyTimer::stop() {
    auto end_time = std::chrono::high_resolution_clock::now();
    return std::chrono::duration_cast<std::chrono::microseconds>(end_time - start_time).count();
}

TradeAnalyzer::TradeAnalyzer(const OrderBook& ob, const std::string& exchange, FeeTier tier)
    : order_book(ob), exchange_name(exchange), fee_tier(tier) {}

    const OrderBook& TradeAnalyzer::getOrderBook() const {
        return order_book;
}

double TradeAnalyzer::computeSpread() const {
    if (order_book.bids.empty() || order_book.asks.empty()) return 0.0;
    return order_book.asks.front().price - order_book.bids.front().price;
}

double TradeAnalyzer::computeVolatility(const std::vector<double>& prices) const {
    if (prices.size() < 2) return 0.0;

    std::vector<double> log_returns;
    for (size_t i = 1; i < prices.size(); ++i)
        log_returns.push_back(std::log(prices[i] / prices[i - 1]));

    double mean = std::accumulate(log_returns.begin(), log_returns.end(), 0.0) / log_returns.size();
    double variance = 0.0;
    for (double r : log_returns)
        variance += (r - mean) * (r - mean);

    variance /= log_returns.size();
    return std::sqrt(variance);
}

double TradeAnalyzer::computeSlippage(
    double usd_amount,
    double volatility,
    double spread,
    const std::string& exchange
) const {
    // Dummy linear regression coefficients for OKX
    const double beta0 = 0.01;   // base slippage
    const double beta1 = 0.12;   // log(order size)
    const double beta2 = 0.6;    // volatility weight
    const double beta3 = 0.25;   // spread weight

    // You could customize coefficients based on exchange if needed
    // if (exchange == "BINANCE") { ... }

    double log_order = std::log(usd_amount);
    return beta0 + beta1 * log_order + beta2 * volatility + beta3 * spread;
}

static double sigmoid(double x) {
    return 1.0 / (1.0 + std::exp(-x));
}

double get_taker_fee_percent(FeeTier tier) {
    switch (tier) {
        case FeeTier::TIER1: return 0.001;
        case FeeTier::TIER2: return 0.0005;
        case FeeTier::TIER3: return 0.0001;
    }
    return 0.001;
}

double TradeAnalyzer::computeFee(double usd_amount) const {
    double fee_percent = get_taker_fee_percent(fee_tier);
    return usd_amount * fee_percent;
}

double TradeAnalyzer::computeMarketImpact(double usd_qty, MarketImpactParams params) const {
    return params.eta * usd_qty + params.gamma * usd_qty * usd_qty;
}

MakerTakerResult TradeAnalyzer::estimateMakerTaker(const MakerTakerInput& input) const {
    // Heuristic weights
    double w_market = 2.0;
    double w_volatility = 1.5;
    double w_depth = -1.0;
    double w_time = -0.5;

    double score = 0.0;
    score += w_market * (input.is_market_order ? 1.0 : 0.0);
    score += w_volatility * input.volatility;
    score += w_depth * input.order_book_depth;
    score += w_time * input.time_since_last_trade;

    // Convert score to probability using sigmoid
    double taker_prob = sigmoid(score);
    double maker_prob = 1.0 - taker_prob;

    return {maker_prob, taker_prob};
}

double TradeAnalyzer::computeNetCost(const NetCostInput& input) const {
    return input.slippage + input.fee + input.market_impact;
}

TradeCostResult TradeAnalyzer::analyze(const OrderBook& ob,
                                       double usd_amount,
                                       FeeTier tier,
                                       const MarketImpactParams& mi_params,
                                       const MakerTakerInput& mk_input,
                                       const std::vector<double>& recent_prices,
                                       const std::string& exchange_name) const {
    double spread = computeSpread();
    double volatility = computeVolatility(recent_prices);
    double slippage = computeSlippage(usd_amount, volatility, spread, exchange_name);
    double fee = computeFee(usd_amount);
    double market_impact = computeMarketImpact(usd_amount, mi_params);

    NetCostInput net_input{usd_amount, slippage, fee, market_impact};
    double net_cost = computeNetCost(net_input);

    MakerTakerResult maker_taker_result = estimateMakerTaker(mk_input);

    return TradeCostResult{
        slippage,
        fee,
        market_impact,
        net_cost,
        maker_taker_result
    };
}
